src/voice/VoiceRecorder.ts

```diff
@@ -60,12 +56,32 @@ export class VoiceRecorder {
             },
         });
         this.recorderContext = new AudioContext({
-            latencyHint: "interactive",
+            // latencyHint: "interactive", // we don't want a latency hint (this causes data smoothing)
             sampleRate: SAMPLE_RATE, // once again, the browser will resample for us
         });
         this.recorderSource = this.recorderContext.createMediaStreamSource(this.recorderStream);
-        this.recorderFreqNode = this.recorderContext.createAnalyser();
-        this.recorderSource.connect(this.recorderFreqNode);
+        this.recorderFFT = this.recorderContext.createAnalyser();
+
+        // Bring the FFT time domain down a bit. The default is 2048, and this must be a power
+        // of two. We use 64 points because we happen to know down the line we need less than
+        // that, but 32 would be too few. Large numbers are not helpful here and do not add
+        // precision: they introduce higher precision outputs of the FFT (frequency data), but
+        // it makes the time domain less than helpful.
+        this.recorderFFT.fftSize = 64;
+
+        // We use an audio processor to get accurate timing information.
+        // The size of the audio buffer largely decides how quickly we push timing/waveform data
+        // out of this class. Smaller buffers mean we update more frequently as we can't hold as
+        // many bytes. Larger buffers mean slower updates. For scale, 1024 gives us about 30Hz of
+        // updates and 2048 gives us about 20Hz. We use 1024 to get as close to perceived realtime
+        // as possible. Must be a power of 2.
+        this.recorderProcessor = this.recorderContext.createScriptProcessor(1024, CHANNELS, CHANNELS);
```

Comment:
Won't this stop working soon? (It's marked as deprecated?)

Reply:
err. where is is deprecated? It's the only thing recommended for this use case (that I could find)

Reply:
apparently I never looked at the giant red banner. 

The replacement API is a bit annoying to work with, but I've done so here: https://github.com/matrix-org/matrix-react-sdk/pull/5801/commits/5c685dcf35fbe8c896f22882a6843dc0cc0b835d